File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/TemplateOrInterface.scala
// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.daml.lf
package language

// glorified Either to handle template or interface cases
sealed abstract class TemplateOrInterface[+T, +I]
object TemplateOrInterface {
  final case class Template[+T](value: T) extends TemplateOrInterface[T, Nothing]
  final case class Interface[+I](value: I) extends TemplateOrInterface[Nothing, I]
}

File: compiler/daml-lf-ast/src/DA/Daml/LF/TemplateOrInterface.hs
-- Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
-- SPDX-License-Identifier: Apache-2.0

module DA.Daml.LF.TemplateOrInterface
  ( TemplateOrInterface (..)
  , TemplateOrInterface'
  ) where

-- | Glorified 'Either' to handle template or interface cases
data TemplateOrInterface a b
  = Template a
  | Interface b

-- | For the common case where both type arguments are the same
type TemplateOrInterface' a = TemplateOrInterface a a

